### PR TITLE
♻️ changed dierctor-v2 -> dy-sidecar API retry policy

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

When attempting a connection from `director-v2` to `dy-sidecar`, the API waited for a maximum 7 seconds (measured). This is due to current tenacity policy which retries with an `exponential_wait` and `stop_after` 3 attempts (1 +2 +4 = 7 seconds)  of total wait time.

The new policy will make it wait `DYNAMIC_SIDECAR_CLIENT_REQUEST_TIMEOUT_S` with an exponential wait before giving up on the request. This will translate in pauses of: 1, 2, 3, 4,  8, 16, 32 seconds

## Related issue/s

- fixes https://github.com/ITISFoundation/osparc-simcore/issues/3478
- https://github.com/ITISFoundation/osparc-issues/issues/638

## Checklist

- [x] Unit tests for the changes exist

### CODE REVIEW

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_client_api_base.py [hunk 1]
```diff
@@ -37,30 +37,42 @@ async def get_retry_for_status(self) -> Response:
...
         assert log_message.startswith("Unexpected error")
-        assert log_message.endswith(f"(attempt [{unexpected_counter}/{retry_count}])")
+        assert log_message.endswith(f"(attempt {unexpected_counter})")
         unexpected_counter += 1
 
 
 @pytest.fixture
-def thick_client() -> FakeThickClient:
-    return FakeThickClient(request_max_retries=1)
+def request_max_network_issues_tolerance_s() -> int:
```
Comment (sanderegg) [inline services/director-v2/tests/unit/test_modules_dynamic_sidecar_client_api_base.py:53]: suggestion: also rename this to ```request_timeout```

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_client_api_public.py [hunk 2]
```diff
@@ -50,7 +50,7 @@ def mock_env(monkeypatch: MonkeyPatch, mock_env: EnvVarsDict) -> None:
     monkeypatch.setenv("POSTGRES_DB", "")
 
     # reduce number of retries to make more reliable
-    monkeypatch.setenv("DYNAMIC_SIDECAR_API_CLIENT_REQUEST_MAX_RETRIES", "1")
+    monkeypatch.setenv("DYNAMIC_SIDECAR_NETWORK_ISSUES_TOLERANCE_S", "3")
```
Comment (sanderegg) [inline services/director-v2/tests/unit/test_modules_dynamic_sidecar_client_api_public.py:53]: suggestion: ```DYNAMIC_SIDECAR_CLIENT_REQUEST_TIMEOUT_S```

File: services/director-v2/tests/unit/test_modules_dynamic_sidecar_client_api_public.py [hunk 3]
```diff
@@ -68,16 +68,20 @@ async def dynamic_sidecar_client(
 
 
 @pytest.fixture
-def retry_count() -> int:
-    return 2
+def request_max_network_issues_tolerance_s() -> int:
```
Comment (sanderegg) [inline services/director-v2/tests/unit/test_modules_dynamic_sidecar_client_api_public.py:71]: same here

**Review by sanderegg — APPROVED** [review]
Ok this will probably already help a bit.
Now what do we do in case of high load? Did you clarify whether you first check if the docker engine is up and running before sending these HTTP requests?
I have been testing a bit with docker events, and I think we should probably make use of that to reduce the stress on the docker engine.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_base.py [hunk 4]
```diff
@@ -68,11 +65,11 @@ async def request_wrapper(zelf: "BaseThinClient", *args, **kwargs) -> Response:
         # pylint: disable=protected-access
         try:
             async for attempt in AsyncRetrying(
-                stop=stop_after_attempt(zelf._request_max_retries),
+                stop=stop_after_delay(zelf._request_max_network_issues_tolerance_s),
```
[the remark this answers is not in the document]
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_base.py:68]: That is correct. Will also rename it internally, as it makes more sense.

**Comment by GitHK** [comment]
> Now what do we do in case of high load? Did you clarify whether you first check if the docker engine is up and running before sending these HTTP requests?
I'm on it now, will be addressed in this issue https://github.com/ITISFoundation/osparc-simcore/issues/3581

> I have been testing a bit with docker events, and I think we should probably make use of that to reduce the stress on the docker engine.
I think that's a great idea, I also had in mind a few things to reduce docker engine stress, we can have a chat about it.

**Review by pcrespov — APPROVED** [review]
It seems this might help what we observed yesterday but IMO we need to find a solution that does not so strongly depend on these timeouts ...We shoulld follow up on this

**Review by GitHK — COMMENTED** [review]
.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_base.py [hunk 5]
```diff
@@ -68,11 +65,11 @@ async def request_wrapper(zelf: "BaseThinClient", *args, **kwargs) -> Response:
         # pylint: disable=protected-access
         try:
             async for attempt in AsyncRetrying(
-                stop=stop_after_attempt(zelf._request_max_retries),
+                stop=stop_after_delay(zelf.request_timeout),
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_base.py:68]: Side effect of this timeout:
- service has an extra (60-7) second delay before starting (there are issues with the logic of the health check event)

Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_base.py:68]: - requires https://github.com/ITISFoundation/osparc-simcore/pull/3586

**Pull request merged**